File: modules/Accounts/TDS Payment/TDS Payment.py
class DocType:
  def __init__(self,d,dl):
    self.doc, self.doclist = d, dl
  
  def autoname(self):
    self.doc.name = make_autoname(self.doc.naming_series+'/.####')

  def get_pan_tan(self):
    comp_det=sql("Select pan_no,tan_no from `tabCompany` where name = '%s'"%(self.doc.company))
    if not comp_det:
      msgprint("No PAN or TAN number is mentioned in comapny")
      ret = {
      'pan_no':  '',
      'tan_no' :  ''
    }
    else:
      ret = {
        'pan_no': cstr(comp_det[0][0]) ,
        'tan_no' : cstr(comp_det[0][1]) 
      }
   
    return cstr(ret)
  
  def get_acc_head(self,args):
    args=eval(args)
    acc_head=sql("Select account_head from `tabTDS Category Account` where parent='%s' and account_type='%s'"%(self.doc.tds_category,args['acc_type']))
    acc_head = acc_head and acc_head[0][0] or ''
    ret = {'acc_head':acc_head, 'tds_value':args['tds_value']}
    return ret
  
  def get_tds_list(self):
    self.doc.clear_table(self.doclist,'tds_payment_details')
    
    self.doc.total_main_tds= 0
    self.doc.total_surcharge= 0
    self.doc.total_edu_cess= 0
    self.doc.total_sh_edu_cess= 0
    self.doc.total_tds= 0
    
    import datetime
    if self.doc.tds_category:
      if not self.doc.from_date or not self.doc.to_date:
        msgprint("Please enter from date and to date")
      else:
        pv_det= sql("Select name,credit_to,grand_total,posting_date from `tabPayable Voucher` where tds_category='%s' And posting_date<= '%s' And posting_date >='%s'  And docstatus=1 Order By posting_date"%(self.doc.tds_category,self.doc.to_date,self.doc.from_date))
        if pv_det:
          self.make_tds_table(pv_det)
        
        jv_det= sql("Select name, supplier_account, total_debit,posting_date from `tabJournal Voucher` where tds_category='%s' And posting_date<= '%s' And posting_date >='%s' And docstatus=1 Order By posting_date"%(self.doc.tds_category,self.doc.to_date,self.doc.from_date))
        if jv_det:
          self.make_tds_table(jv_det)
          
    else:
      msgprint("Please select tds category")

  def make_tds_table(self,det_touple):
   
    idx=1
    for v in det_touple:
      v_no=v and v[0] or ''
      if not sql("select name from `tabTDS Payment Detail` where voucher_no = '%s' and parent != '%s' and docstatus = 1" % (v_no, self.doc.name)):
        child = addchild(self.doc, 'tds_payment_details', 'TDS Payment Detail', 1, self.doclist)
        child.voucher_no = v_no
        child.party_name = v and v[1] or ''
        child.amount_paid = v and flt(v[2]) or ''
        child.date_of_payment =v and v[3].strftime('%Y-%m-%d') or ''
        
        tds_det = sql("Select `tabPV Ded Tax Detail`.tds_type,`tabPV Ded Tax Detail`.ded_amount from `tabPV Ded Tax Detail` where `tabPV Ded Tax Detail`.parent='%s'"%(v_no))
        
        for d in tds_det:
          
          if d[0]=='Main':
            child.tds_main=flt(d[1])
          elif d[0]=='Surcharge':
            child.surcharge=flt(d[1])
          elif d[0]=='Edu Cess':
            child.edu_cess=flt(d[1])
          elif d[0]=='SH Edu Cess':
            child.sh_edu_cess=flt(d[1])
        
        child.tds_main=child.tds_main
        child.surcharge=child.surcharge
        child.edu_cess=child.edu_cess
        child.sh_edu_cess=child.sh_edu_cess      
        child.total_tax_amount=flt(child.tds_main)+flt(child.surcharge)+flt(child.edu_cess)+flt(child.sh_edu_cess)
        child.idx=idx
        idx=idx+1
      
        self.doc.total_main_tds= flt(self.doc.total_main_tds)+flt(child.tds_main)
        self.doc.total_surcharge= flt(self.doc.total_surcharge)+flt(child.surcharge)
        self.doc.total_edu_cess= flt(self.doc.total_edu_cess)+flt(child.edu_cess)
        self.doc.total_sh_edu_cess= flt(self.doc.total_sh_edu_cess)+flt(child.sh_edu_cess)
        self.doc.total_tds= flt(self.doc.total_tds)+flt(child.total_tax_amount)

  def validate(self):
    if self.doc.amended_from and not self.doc.amendment_date:
      msgprint("Please Enter Amendment Date")
      raise Exception
  
  def check_bsr_code(self):
    if not self.doc.bsr_code or not self.doc.date_of_receipt or not self.doc.challan_no:
      msgprint("Voucher can not be submitted without Cheque/DD No and BSR Code and Date of Receipt and Challan ID No")
      raise Exception
  
  def on_submit(self):
    self.check_bsr_code()

File: modules/Accounts/TDS Payment/TDS Payment.js
cur_frm.cscript.onload = function(doc,cdt,cdn){
  if(doc.company)get_server_fields('get_pan_tan','','',doc,cdt,cdn,1);
  if (doc.docstatus==0) hide_field('Make Bank Voucher');
}


cur_frm.cscript.company = function(doc,cdt,cdn){
  if(doc.company)get_server_fields('get_pan_tan','','',doc,cdt,cdn);
}

cur_frm.cscript.to_date = function(doc,cdt,cdn){
  if(doc.from_date>doc.to_date){
    alert("From date can not be greater than To date");
    doc.to_date='';
    refresh_field('to_date');
}
}

cur_frm.cscript.tds_category = function(doc,cdt,cdn){
  if (doc.tds_category){
    var tds_acc_type = new Array();
    tds_acc_type=['Main','Surcharge','Edu Cess','Sh Edu Cess'];
    for(i=0;i<tds_acc_type.length;i++){
      args={'acc_type':tds_acc_type[i]}
      
      $c('runserverobj', args={'method':'get_acc_head', arg:docstring(args),'docs':compress_doclist([doc])}, 
        function(r, rt) {
          pscript.acc_head = r.message;
        }
      );
      if (tds_acc_type[i]=='Main') doc.main_acc = pscript.acc_head;
      else if (tds_acc_type[i]=='Surcharge') doc.surcharge_acc=pscript.acc_head;
      else if (tds_acc_type[i]=='Edu Cess') doc.edu_cess_acc=pscript.acc_head;
      else if (tds_acc_type[i]=='Sh Edu Cess') doc.sh_edu_cess_acc=pscript.acc_head;
    }
    alert(doc.main_acc);
    alert(doc.surcharge_acc);
    alert(doc.edu_cess_acc);
    alert(doc.sh_edu_cess_acc);
    refresh_field('main_acc');
    refresh_field('surcharge_acc');
    refresh_field('edu_cess_acc');
    refresh_field('sh_edu_cess_acc');
  }
}

cur_frm.cscript.total_tds = function(doc,cdt,cdn){
  var tot_main=0,tot_surcharge=0,tot_edu=0,tot_sh_edu=0,tot_tax=0;
  var tds_l = getchildren('TDS Payment Detail',doc.name,'tds_payment_details')
  for(var i=0; i<tds_l.length;i++){
    tot_main = flt(tot_main)+flt(tds_l[i].tds_main);
    tot_surcharge = flt(tot_surcharge)+flt(tds_l[i].surcharge);
    tot_edu = flt(tot_edu)+flt(tds_l[i].edu_cess);
    tot_sh_edu = flt(tot_sh_edu)+flt(tds_l[i].sh_edu_cess);
    tot_tax = flt(tot_tax)+flt(tds_l[i].total_tax_amount);
  }
  var tds_values = new Array();
  
  tds_values['Main'] = tot_main;
  tds_values['Surcharge'] = tot_surcharge;
  tds_values['Edu Cess'] = tot_edu;
  tds_values['Sh Edu Cess'] = tot_sh_edu;
  tds_values['TotalTDS'] = tot_tax;
  
  return tds_values;
}

// Make Journal Voucher
// --------------------

cur_frm.cscript['Make Bank Voucher'] = function(doc, dt, dn) {
  
  var call_back = function(r,rt) {
    cur_frm.cscript.make_jv(doc,dt,dn,r.message);
  }
  args = {'credit_to':'', 'company':doc.company};
  $c_obj('ERP Setup', 'get_bank_defaults', docstring(args), call_back);

}

cur_frm.cscript.make_jv = function(doc, dt, dn, det) {

  tds_values=cur_frm.cscript.total_tds(doc,cdt,cdn);
  
  
  var jv = LocalDB.create('Journal Voucher');
  jv = locals['Journal Voucher'][jv];
  jv.voucher_type = det.def_bv_type;
  jv.voucher_series = det.def_bv_series;
  jv.voucher_date = doc.transaction_date;
  jv.posting_date = doc.posting_date;
  jv.remark = repl('Payment against voucher %(vn)s for %(rem)s', {vn:doc.name, rem:doc.remarks});
  jv.total_debit = tds_values['TotalTDS'];
  jv.total_credit = tds_values['TotalTDS'];
  jv.fiscal_year = doc.fiscal_year;
  jv.company = doc.company;
  
  // debit to creditor
  
  var callback_tds = function(tds_value, credit_to){
    //pscript.credit_to=r.message;
    pscript.credit_to=credit_to;
    var d1 = LocalDB.add_child(jv, 'Journal Voucher Detail', 'entries');
    d1.account = pscript.credit_to;
    
    //d1.debit = tds_values[i];
    d1.debit = tds_value;
    //d1.balance = det.acc_balance;
    //d1.against_voucher = doc.name;
  }

  for (var i in tds_values){
    if (tds_values[i]>0 && i!= 'TotalTDS'){
      args={'acc_type':i, 'tds_value':tds_values[i]};
      
      $c('runserverobj', args={'method':'get_acc_head', arg:docstring(args), 'docs':compress_doclist([doc])}, 
        function(r, rt) {  
          callback_tds(r.message.tds_value, r.message.acc_head);
        }
      );
    }
  }
  // credit to bank
  j='TotalTDS'
  var d1 = LocalDB.add_child(jv, 'Journal Voucher Detail', 'entries');
  d1.account = det.def_bank_account;
  d1.balance = det.bank_balance;
  d1.credit = tds_values[j];
  
  loaddoc('Journal Voucher', jv.name);
  //alert("journal voucher made");
}

// Refresh
// -------

cur_frm.cscript.refresh = function(doc, dt, dn) {
  // Show / Hide button
  if(doc.docstatus==1) { unhide_field('Make Bank Voucher'); }
  else hide_field('Make Bank Voucher');
}